rfcs/006-package-management.md

```diff
@@ -0,0 +1,560 @@
+a shorthand?
+
+### Alternative: manifest file name
+
+Bikeshed the name "electroplate.ncl".
+
+## Import statements
+
+The manifest file assigns a name to each dependency; to import the dependency
+named `foo` you simply write `import foo`. That is, an `import` statement either
+takes a string in quotes -- in which case it imports a path -- or an identifier
+without quote -- in which case it imports a package.
+
+The `import foo` expression evaluates to the contents of `main.ncl` in `foo`'s
+root directory.
+
+### Alternative: specify the entrypoint in the manifest
+
+Since the package manifest is a nickel file, it could contain an arbitrary nickel
+expression as its entrypoint. That is, the `import foo` expression could evaluate
+to `(import "<path-to-foo>/electroplate.ncl").entrypoint`, and we could provide
+a default `entrypoint | default = import "main.ncl"` manifest entry. It might
+require some care to avoid looking for the "main.ncl" file if it doesn't exist.
+
+### Alternative: other entry points
+
+We've hardcoded `main.ncl` as the entry point of every package, but what if
+they want to expose multiple entry points? For example, node allows a package's
+manifest to specify the entry point(s). This is probably not very important to
+support, as you can just put
+
+```nickel
+{
+  other = import "./other.ncl",
+  blah = import "./blah.ncl",
+}
+```
+
+in your package's `main.ncl`, to provide "other" and "blah" as other entry points.
+
+### Alternative: entry points are top-level files
+
+Instead of hardcoding `main.ncl`, we could say that every file in the package's
+top-level directory is publicly accessible. Package authors could keep implementation
+details private by putting code in subdirectories. The import syntax would
+need to change somehow to specify the entry point.
+
+### Alternative: namespace the import tool
+
+Our initial intention for packaging was to allow for the usage of multiple
+different package management tools. This RFC only proposes one such tool, but
+maybe the import syntax could be designed with other tools in mind. For example,
+it could be `import foo from electroplate` with the idea that future nickel
+versions might add, say, `import foo from nix-flake`.
+
+One problem with this approach is that even if you switch package managers, you
+still want to import the same packages. So probably it doesn't make sense to
+include the name of the package manager in the nickel source.
+
+## CLI interface
+
+We will build package management straight into the nickel CLI. `nickel eval`,
+`nickel export`,
+and similar commands will do an additional package-management step before the
+actual evaluation. They will start by searching for an `electroplate.ncl` file.
+If one is found, we will evaluate it. We will then search for a lock-file.
+If one is found, it will be used to guide dependency resolution; if not, we will
+do dependency resolution from scratch and write out the generated lock-file.
```

Comment:
I wonder what's the cost of doing that on each evaluation. As opposed to compiled languages, we might have to pay the dependency resolution step on each run, including for small-ish script (as opposed to say `cargo` when you pay the price at build time but not at run time). Do you think that would be an argument in favor of better separated package-management, where `nickel eval` would just pick the existing lockfile or fail but not perform the update automatically? Or at least an argument for not updating the lockfile automatically? Or do you expect those costs to be negligible?

Reply:
Good question, I guess we'd have to measure it to be sure. I expect version resolution to be quick, especially if there's an up-to-date lockfile but I don't actually have data.

Having to manually trigger a re-resolution is definitely an option; I think poetry and npm do this. I personally find it a bit annoying, as it's an extra step that's easy to forget.